### PR TITLE
resolves #1490 consistent SVG height and width between diagrams

### LINKED ISSUE #1490: SVG width and height attributes inconsistencies between diagrams
**Is your feature request related to a problem? Please describe.**

The width and height attributes are inconsistent between diagrams. All examples below are taken from the live editor: https://mermaid-js.github.io/mermaid-live-editor (select your diagram type and then click on "Link to SVG").

<ins>Flowchart</ins>

```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="312" height="445.6000061035156" viewBox="0 0 312 445.6000061035156">
```
Both width and height are absolute values :+1:

<ins>Sequence diagram</ins>
```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="100%" height="100%" style="max-width:516px;" viewBox="-50 -10 516 301">
```

100% height and width with an inline style to set a maximum width of 516px. Please note that the height is 301px (the value is present in the `viewBox` attribute).

<ins>Class diagram</ins>
```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="393" height="236" viewBox="-20 -20 393 236">
```

Both width and height are absolute values :+1:

<ins>State diagram</ins>

```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="192.5" viewBox="-26.75  0 110 266">
```

The width is defined but the height is not. The height and width can be inferred from the `viewBox` attribute: 110px width and 266px height. Please note that the width is not equals to the value of the viewBox.

<ins>Gantt diagram</ins>

```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="100%" height="100%" viewBox="0 0 800 196">
```

100% height and width. The height and width can be inferred from the `viewBox` attribute: 800px width and 196px height.

<ins>Pie chart</ins>

```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="100%" height="100%" viewBox="0 0 800 450">
```

100% height and width. The height and width can be inferred from the `viewBox` attribute: 800px width and 450px height.

<ins>ER diagram</ins>

```xml
<svg xmlns="http://www.w3.org/2000/svg" id="mermaid-svg" width="100%" height="640" style="max-width: 769px;" viewBox="-118 0 769 640">
```

100% height with a fixed height of 640px and an inline style to set a maximum width of 769px.

**Describe the solution you'd like**

To be consistent, I think that all diagrams should declare the same attributes.
I believe that it's a good practice to set both the width and height with absolute values (i.e. the initial size with the right ratio). Then you can use CSS or inline styles to adjust the SVG to the container in your page.

Here's an example with the class diagram:

| Initial size | Fit container
| -- | -- |
| init | height500px-widht100 |

**Describe alternatives you've considered**

I guess we could also set the width and height to 100% and if someone needs an initial fixed size then the values can be inferred from the `viewBox`.
Having said that, setting width and height to 100% does not give the user any additional values. If the user wants to fill all the space available in the container then (s)he can easily set width and height to 100% in CSS:

```css
#mermaid-svg {
  width: 100%;
  height: 100%;
}
```

On the other hand, extracting the initial fixed size from the `viewBox` is much harder and, as far as I know, you cannot do it using only CSS.

### PR DESCRIPTION
**Pull request by Mogztter**
## :bookmark_tabs: Summary

Use the same width and height attributes on all diagrams depending on the `useMaxWidth` option.
By default, this option is `true`.

Please note that some diagrams have an (optional) option named `useWidth` to control the value of the width.

Resolves #1490

## :straight_ruler: Design Decisions

I've declared utility functions in `utils.js` to avoid code duplication.

### :clipboard: Tasks

Make sure you
- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md) 
- [x] :computer: have added unit/e2e tests
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by Mogztter** [comment]
The tests are failing because the size of the generated diagram is slighty different on Travis compared to what I get on my local machine:

```
CypressError: Timed out retrying: expected '<svg#mermaid-1598213404098>' to have attribute 'style' with the value 'max-width: 162.28125px;', but the value was 'max-width: 160.28125px;'
```

Not sure how to work around this issue... Maybe I should just check that the attribute contains/starts with `'max-width: 16`? What do you think?

I don't think the interaction failures are related? The error message is:

```
CypressError: Timed out retrying: expected 'http://localhost:9000/click_security_loose.html' to equal 'http://localhost:9000/webpackUsage.html'
```

**Comment by Mogztter** [comment]
> I don't think the interaction failures are related? The error message is:
> CypressError: Timed out retrying: expected 'http://localhost:9000/click_security_loose.html' to equal 'http://localhost:9000/webpackUsage.html'

The error message is misleading but it was indeed related to my changes.
I've added e2e tests on all diagrams to make sure that my code works as expected.

In the process, I found and fixed 2 bugs on pie charts :tada:

File: src/diagrams/pie/pieRenderer.js [hunk 1]
```diff
@@ -31,34 +33,31 @@ export const draw = (txt, id) => {
...
-    var width = w; // 450
-    var height = 450;
     var margin = 40;
     var legendRectSize = 18;
     var legendSpacing = 4;
 
     var radius = Math.min(width, height) / 2 - margin;
 
-    var svg = select('#' + id)
-      .append('svg')
```
Comment (Mogztter) [inline src/diagrams/pie/pieRenderer.js:59]: The following code was creating a nested `<svg>` element but I think it was a mistake (as we don't need it)

File: src/diagrams/pie/pieRenderer.js [hunk 2]
```diff
@@ -67,9 +66,8 @@ export const draw = (txt, id) => {
     Object.keys(data).forEach(function(key) {
       sum += data[key];
     });
-    logger.info(data);
```
Comment (Mogztter) [inline src/diagrams/pie/pieRenderer.js:70]: Remove the logger

File: src/mermaidAPI.js [hunk 3]
```diff
@@ -385,7 +385,7 @@ const render = function(id, _txt, cb, container) {
         break;
       case 'pie':
         cnf.class.arrowMarkerAbsolute = cnf.arrowMarkerAbsolute;
-        pieRenderer.setConf(cnf.class);
```
Comment (Mogztter) [inline src/mermaidAPI.js:388]: Fixed a bug, the configuration for pie charts was ignored (i.e. never assigned)

**Comment by knsv** [comment]
Thanks! Making this consistent is the way to go!

**Comment by knsv** [comment]
Thanks @Mogztter for your efforts and thanks @getaaron for your review.

After completed rendering tests it looks really good apart from this case where there is a very large flowchart as per the image below. Here we see a changed behavior.

image

**Comment by knsv** [comment]
I can not reproduce this locally so it seems it is triggered by something during the CI build.

**Comment by Mogztter** [comment]
Could you please share the command you are using to reproduce this issue locally?

The image on the right is the expected result and the image on the left is the result we get with my changes?

**Comment by knsv** [comment]
Se comment above. Good job!

**Comment by Mogztter** [comment]
> Se comment above. Good job!

Perfect 👌

**Pull request merged**